staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"math"
+
+	"github.com/google/cel-go/checker"
+	"github.com/google/cel-go/common"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+)
+
+// CostEstimator implements CEL's interpretable.ActualCostEstimator and checker.CostEstimator.
+type CostEstimator struct {
+	// SizeEstimator provides a CostEstimator.EstimateSize that this CostEstimator will delegate size estimation
+	// calculations to if the size is not well known (i.e. a constant).
+	SizeEstimator checker.CostEstimator
+}
+
+func (l *CostEstimator) CallCost(function, overloadId string, args []ref.Val, result ref.Val) *uint64 {
+	switch function {
+	case "isSorted", "sum", "max", "min", "indexOf", "lastIndexOf":
+		var cost uint64
+		if lister, ok := args[0].(traits.Lister); ok {
+			// Start with a base 1 cost for comparing each element in the list.
+			size := uint64(lister.Size().(types.Int))
```

Comment:
are all of these casts guaranteed safe?

Reply:
Yes, the size ones are safe. xref from cel-go: https://github.com/google/cel-go/blob/9e07c9972efa837cb00dda458286bdf96eb15309/common/types/map.go#L650

cc @TristonianJones (incase there is any reason I should guard the check that is super subtle here).

Reply:
You're not missing anything subtle here. The type of `Size` should be a `types.Int`. I'm trying to recall why I structured the return value that way and I'm having some trouble recalling because it was quite a while ago.